<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.56">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Paulo Barros">

<title>Leitura de arquivos Illumina com snpStats – Introdução a Estatística Genômica</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./index.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./1_snpStats.html">Leitura de arquivos Illumina com snpStats</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-center sidebar-header">
      <a href="./index.html" class="sidebar-logo-link">
      <img src="./images/LogoStatsGen.png" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
      <div class="sidebar-tools-main tools-wide">
    <a href="https://paulobarros.com.br" title="paulobarros.com.br" class="quarto-navigation-tool px-1" aria-label="paulobarros.com.br"><i class="bi bi-globe"></i></a>
    <a href="https://github.com/paulobarros/intro-stats-gen" title="Introdução a Estatística Genômica" class="quarto-navigation-tool px-1" aria-label="Introdução a Estatística Genômica"><i class="bi bi-github"></i></a>
    <a href="https://twitter.com/pbarrosjr" title="Paulo Barros Twitter" class="quarto-navigation-tool px-1" aria-label="Paulo Barros Twitter"><i class="bi bi-twitter"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bem-vind@s!</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./1_snpStats.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Leitura de arquivos Illumina com snpStats</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#arquivos-final-report-e-de-mapa" id="toc-arquivos-final-report-e-de-mapa" class="nav-link active" data-scroll-target="#arquivos-final-report-e-de-mapa"><span class="header-section-number">1</span> Arquivos <em>Final Report</em> e de Mapa</a>
  <ul class="collapse">
  <li><a href="#explorando-um-final-report" id="toc-explorando-um-final-report" class="nav-link" data-scroll-target="#explorando-um-final-report"><span class="header-section-number">1.1</span> Explorando um <em>Final Report</em></a></li>
  <li><a href="#arquivo-de-mapa" id="toc-arquivo-de-mapa" class="nav-link" data-scroll-target="#arquivo-de-mapa"><span class="header-section-number">1.2</span> Arquivo de Mapa</a></li>
  </ul></li>
  <li><a href="#fazendo-a-leitura-dos-genótipos-com-o-snpstats" id="toc-fazendo-a-leitura-dos-genótipos-com-o-snpstats" class="nav-link" data-scroll-target="#fazendo-a-leitura-dos-genótipos-com-o-snpstats"><span class="header-section-number">2</span> Fazendo a leitura dos genótipos com o <code>snpStats</code></a></li>
  <li><a href="#snpstats" id="toc-snpstats" class="nav-link" data-scroll-target="#snpstats"><span class="header-section-number">3</span> <code>snpStats</code></a>
  <ul class="collapse">
  <li><a href="#criando-a-lista-de-final-reports" id="toc-criando-a-lista-de-final-reports" class="nav-link" data-scroll-target="#criando-a-lista-de-final-reports"><span class="header-section-number">3.1</span> Criando a lista de <em>Final Reports</em></a></li>
  <li><a href="#criando-os-vetores-de-id-dos-snps" id="toc-criando-os-vetores-de-id-dos-snps" class="nav-link" data-scroll-target="#criando-os-vetores-de-id-dos-snps"><span class="header-section-number">3.2</span> Criando os vetores de ID dos SNPs</a></li>
  <li><a href="#criando-os-vetores-de-id-dos-animais" id="toc-criando-os-vetores-de-id-dos-animais" class="nav-link" data-scroll-target="#criando-os-vetores-de-id-dos-animais"><span class="header-section-number">3.3</span> Criando os vetores de ID dos Animais</a></li>
  <li><a href="#gerando-o-arquivo-de-genótipos" id="toc-gerando-o-arquivo-de-genótipos" class="nav-link" data-scroll-target="#gerando-o-arquivo-de-genótipos"><span class="header-section-number">3.4</span> Gerando o arquivo de genótipos</a></li>
  <li><a href="#explorando-o-objeto-de-genótipos-genodat" id="toc-explorando-o-objeto-de-genótipos-genodat" class="nav-link" data-scroll-target="#explorando-o-objeto-de-genótipos-genodat"><span class="header-section-number">3.5</span> Explorando o objeto de genótipos <code>genodat</code></a></li>
  </ul></li>
  <li><a href="#salvando-o-arquivo-de-genótipos" id="toc-salvando-o-arquivo-de-genótipos" class="nav-link" data-scroll-target="#salvando-o-arquivo-de-genótipos"><span class="header-section-number">4</span> Salvando o arquivo de genótipos</a>
  <ul class="collapse">
  <li><a href="#salvando-.rdata" id="toc-salvando-.rdata" class="nav-link" data-scroll-target="#salvando-.rdata"><span class="header-section-number">4.1</span> Salvando .Rdata</a></li>
  <li><a href="#salvando-em-formato-snpmatrix" id="toc-salvando-em-formato-snpmatrix" class="nav-link" data-scroll-target="#salvando-em-formato-snpmatrix"><span class="header-section-number">4.2</span> Salvando em formato SNPMatrix</a></li>
  <li><a href="#explorando-nosso-arquivo-genodat.txt" id="toc-explorando-nosso-arquivo-genodat.txt" class="nav-link" data-scroll-target="#explorando-nosso-arquivo-genodat.txt"><span class="header-section-number">4.3</span> Explorando nosso arquivo <code>genodat.txt</code></a></li>
  </ul></li>
  <li><a href="#código-completo" id="toc-código-completo" class="nav-link" data-scroll-target="#código-completo"><span class="header-section-number">5</span> Código Completo</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/paulobarros/intro-stats-gen/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div><div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="1_snpStats.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Leitura de arquivos Illumina com snpStats</h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliation</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Paulo Barros <a href="mailto:pbarrosbio@gmail.com" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0001-9839-0796" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            PPZ - UESB - GACOM
          </p>
      </div>
  </div>

<div class="quarto-title-meta">

      
  
    
  </div>
  


</header>


<section id="arquivos-final-report-e-de-mapa" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="arquivos-final-report-e-de-mapa"><span class="header-section-number">1</span> Arquivos <em>Final Report</em> e de Mapa</h2>
<p>Nesta prática vamos trabalhar com dados de sequencias de um chip comercial de galinhas da <em>Illumina</em>. Um arquivo <strong><em>Final Report</em></strong> é um arquivo de texto gerado a partir do Genome Studio contendo as informações de genótipos dos indivíduos.</p>
<p>O <em>Final Report</em> pode conter diversos campos com informações sobre a genotipagem, aqui vamos cobrir alguns mais importantes:</p>
<table class="table-striped table-hover caption-top table">
<caption>Campos de um <em>Final Report</em></caption>
<colgroup>
<col style="width: 40%">
<col style="width: 60%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Campo</th>
<th style="text-align: left;">Descrição</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Sample Name</td>
<td style="text-align: left;">Nome da amostra</td>
</tr>
<tr class="even">
<td style="text-align: left;">SNP Name</td>
<td style="text-align: left;">Identificador do SNP</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Allele1 Top <br> Allele2 Top</td>
<td style="text-align: left;">Alelos 1 e 2 em nucleotideos correspondentes aos Alelos A e B, reportados na fita Top</td>
</tr>
<tr class="even">
<td style="text-align: left;">Sample ID</td>
<td style="text-align: left;">Identificador da Amostra</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GC Score</td>
<td style="text-align: left;">Medida de qualidade calculada para cada genótipo, varia de 0 a 1</td>
</tr>
<tr class="even">
<td style="text-align: left;">Allele1 Forward <br> Allele2 Forward</td>
<td style="text-align: left;">Alelos 1 e 2 em nucleotideos correspondentes aos Alelos A e B, reportados na fita Forward</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Allele1 AB <br> Allele2 AB</td>
<td style="text-align: left;">Alelos 1 e 2 em genótipos AB</td>
</tr>
</tbody>
</table>
<p>Para consultar todos os campos disponíveis no <em>Final Report</em> você pode visitar o <a href="https://knowledge.illumina.com/microarray/general/microarray-general-reference_material-list/000002861" target="_blank">site da Illumina.</a></p>
<section id="explorando-um-final-report" class="level3" data-number="1.1">
<h3 data-number="1.1" class="anchored" data-anchor-id="explorando-um-final-report"><span class="header-section-number">1.1</span> Explorando um <em>Final Report</em></h3>
<p>Abaixo você pode ver algumas linhas iniciais de um dos arquivos <em>Final Report</em> que serão trabalhos nesta prática.</p>
<div class="sourceCode" id="annotated-cell-1"><pre class="sourceCode numberSource code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode"><a class="code-annotation-anchor" data-target-cell="annotated-cell-1" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-1-1" class="code-annotation-target"><a href="#annotated-cell-1-1"></a>[Header]</span>
<span id="annotated-cell-1-2"><a href="#annotated-cell-1-2"></a>GSGT Version    2.0.5</span>
<span id="annotated-cell-1-3"><a href="#annotated-cell-1-3"></a>Processing Date 10/19/2020 12:00 PM</span>
<span id="annotated-cell-1-4"><a href="#annotated-cell-1-4"></a>Content     Chicken_50K_CobbCons_15000986_A.bpm</span>
<span id="annotated-cell-1-5"><a href="#annotated-cell-1-5"></a>Num SNPs    57636</span>
<span id="annotated-cell-1-6"><a href="#annotated-cell-1-6"></a>Total SNPs  57636</span>
<span id="annotated-cell-1-7"><a href="#annotated-cell-1-7"></a>Num Samples 192</span>
<span id="annotated-cell-1-8"><a href="#annotated-cell-1-8"></a>Total Samples   192</span>
<span id="annotated-cell-1-9"><a href="#annotated-cell-1-9"></a>File    1 of 192</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-1" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-1-10" class="code-annotation-target"><a href="#annotated-cell-1-10"></a>[Data]</span>
<span id="annotated-cell-1-11"><a href="#annotated-cell-1-11"></a>SNP Name    Sample ID   Allele1 - AB    Allele2 - AB    GC Score</span>
<span id="annotated-cell-1-12"><a href="#annotated-cell-1-12"></a>10080_COIII 203900190022_R01C01 B   B   0.6976</span>
<span id="annotated-cell-1-13"><a href="#annotated-cell-1-13"></a>10668_COIII 203900190022_R01C01 B   B   0.7118</span>
<span id="annotated-cell-1-14"><a href="#annotated-cell-1-14"></a>12187_ND4   203900190022_R01C01 B   B   0.6915</span>
<span id="annotated-cell-1-15"><a href="#annotated-cell-1-15"></a>15466_CYTB_MT   203900190022_R01C01 B   B   0.3856</span>
<span id="annotated-cell-1-16"><a href="#annotated-cell-1-16"></a>16262_ND6   203900190022_R01C01 B   B   0.8686</span>
<span id="annotated-cell-1-17"><a href="#annotated-cell-1-17"></a>16370_ND6   203900190022_R01C01 B   B   0.8878</span>
<span id="annotated-cell-1-18"><a href="#annotated-cell-1-18"></a>5831_ND2_MT 203900190022_R01C01 A   A   0.7214</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-1" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-1" data-code-lines="1" data-code-annotation="1"><code>[Header]</code>: As primeiras linhas de um <em>Final Report</em> nos trazem informações sobre o processo de genotipagem. Nesta seção podemos encontrar o número de SNPs, de amostras, e se o arquivo é um arquivo único ou individual (um <em>Final Report</em> por amostra).</span>
</dd>
<dt data-target-cell="annotated-cell-1" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-1" data-code-lines="10" data-code-annotation="2"><code>[Data]</code>: Na seção data inicia o arquivo de fato, é possível então identificar quais colunas estão disponíveis.</span>
</dd>
</dl>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Atenção!
</div>
</div>
<div class="callout-body-container callout-body">
<p>Nem sempre os <em>Final Reports</em> apresentam todos os campos necessários, para fazer a edição é necessário então ter acesso ao projeto original da genotipagem, e fazer uma nova edição para gerar os arquivos no Genome Studio da Illumina. Isso em cenários reais pode variar bastante entre empresas e projetos.</p>
</div>
</div>
</section>
<section id="arquivo-de-mapa" class="level3" data-number="1.2">
<h3 data-number="1.2" class="anchored" data-anchor-id="arquivo-de-mapa"><span class="header-section-number">1.2</span> Arquivo de Mapa</h3>
<p>É no arquivo de Mapas que encontramos as informações de identificação e da posição física de cada SNP no genoma.</p>
<table class="table-striped table-hover caption-top table">
<caption>Arquivo de Mapa</caption>
<colgroup>
<col style="width: 40%">
<col style="width: 60%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Campo</th>
<th style="text-align: left;">Descrição</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Name</td>
<td style="text-align: left;">Identificador do SNP</td>
</tr>
<tr class="even">
<td style="text-align: left;">Chr</td>
<td style="text-align: left;">Cromossomo</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Position</td>
<td style="text-align: left;">Posição no genoma em pares de base</td>
</tr>
<tr class="even">
<td style="text-align: left;">GenTrain Score</td>
<td style="text-align: left;">Métrica de qualidade do cluster de SNP. É o escore do SNP no algoritmo de clusterização GenTrain</td>
</tr>
</tbody>
</table>
<p>Abaixo você pode ver algumas linhas iniciais do arquivo de mapas que utilizaremos. Em um arquivo de mapas, cada SNP ocupa uma linha do arquivo.</p>
<pre><code>Name    Chr Position    GenTrain Score
10080_COIII 0   0   0.7310036
10668_COIII 0   0   0.7387226
12187_ND4   0   0   0.7660004
15466_CYTB_MT   0   0   0.799299
16262_ND6   0   0   0.8815182
16370_ND6   0   0   0.8978758
5831_ND2_MT 0   0   0.7440125
8472_COXII_MT   0   0   0.7310036
Gga_rs10721746  23  5869156 0.873894
Gga_rs10721817  1   107404689   0.8605233 
</code></pre>
</section>
</section>
<section id="fazendo-a-leitura-dos-genótipos-com-o-snpstats" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="fazendo-a-leitura-dos-genótipos-com-o-snpstats"><span class="header-section-number">2</span> Fazendo a leitura dos genótipos com o <code>snpStats</code></h2>
<blockquote class="blockquote">
<p><strong>O nosso objetivo é fazer a leitura dos <em>Final Reports</em> para cada indivíduo e organizá-los em um único arquivo de genótipo para futuras análises.</strong></p>
</blockquote>
</section>
<section id="snpstats" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="snpstats"><span class="header-section-number">3</span> <code>snpStats</code></h2>
<p>Para esta prática vamos precisar dos pacotes <code>BiocManager</code> e <code>snpStats</code>. Você pode fazer a instalação rodando o código abaixo:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb2-2"><a href="#cb2-2"></a>  <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb2-3"><a href="#cb2-3"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"snpStats"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Após a instalação podemos então carregar os pacotes necessários:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="fu">library</span>(snpStats)</span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>De maneira geral, <strong>em um arquivo de genótipos cada linha deve representar um indivíduo</strong> e conter além de <strong>sua identificação</strong>, as <strong>colunas contendo os genótipos para cada SNP</strong>. A depender do software utilizado para as análises genômicas, a configuração do arquivo de genótipo pode envolver mais campos e a codificação dos genótipos também pode variar.</p>
<p>Nesta prática, vamos construir o arquivo de genótipos em três passos:</p>
<ol type="1">
<li><p>Criar uma lista com os nomes de todos os <em>Final Reports</em> pra leitura.</p></li>
<li><p>Criar vetores que contenham o ID dos Animais e o ID dos SNPs.</p></li>
<li><p>Usar o pacote <code>snpStats</code> para gerar o arquivo de genótipos.</p></li>
</ol>
<section id="criando-a-lista-de-final-reports" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="criando-a-lista-de-final-reports"><span class="header-section-number">3.1</span> Criando a lista de <em>Final Reports</em></h3>
<p>Os arquivos para esta prática estão na pasta <code>data/01_snpStats</code>. Vamos usar a função <code>list.files</code> para gerar nossa lista de arquivos a serem lidos.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>filenames <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="at">path =</span> <span class="st">"data/01_snpStats/"</span>,</span>
<span id="cb4-2"><a href="#cb4-2"></a>           <span class="at">pattern =</span> <span class="st">"Final_Report_.*.txt"</span>,</span>
<span id="cb4-3"><a href="#cb4-3"></a>           <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="criando-os-vetores-de-id-dos-snps" class="level3" data-number="3.2">
<h3 data-number="3.2" class="anchored" data-anchor-id="criando-os-vetores-de-id-dos-snps"><span class="header-section-number">3.2</span> Criando os vetores de ID dos SNPs</h3>
<p>Aqui é bem simples, faremos a leitura do arquivo de mapas e vamos extrair a primeira coluna que é a de ID dos SNPs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>snpids <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"data/01_snpStats/SNP-Table-Chickens50k.txt"</span>,</span>
<span id="cb5-2"><a href="#cb5-2"></a>                     <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,</span>
<span id="cb5-3"><a href="#cb5-3"></a>                     <span class="at">header =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<span id="cb5-4"><a href="#cb5-4"></a>          <span class="fu">pull</span>(Name)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="criando-os-vetores-de-id-dos-animais" class="level3" data-number="3.3">
<h3 data-number="3.3" class="anchored" data-anchor-id="criando-os-vetores-de-id-dos-animais"><span class="header-section-number">3.3</span> Criando os vetores de ID dos Animais</h3>
<p>Aqui vamos utilizar a função <code>map</code> do pacote <code>purrr</code> para iterar sobre todos os nossos <em>Final Reports</em> e capturar somente a informação de ID dos animais.</p>
<div class="cell">
<div class="sourceCode cell-code" id="annotated-cell-7"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-7" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-7-1" class="code-annotation-target"><a href="#annotated-cell-7-1"></a>animids<span class="ot">&lt;-</span> filenames <span class="sc">|&gt;</span></span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-7" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-7-2" class="code-annotation-target"><a href="#annotated-cell-7-2"></a>            <span class="fu">map</span>(\(x){</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-7" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-7-3" class="code-annotation-target"><a href="#annotated-cell-7-3"></a>              <span class="fu">scan</span>(x,<span class="at">skip=</span><span class="dv">11</span>,<span class="at">nlines=</span><span class="dv">1</span>,<span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,<span class="at">what=</span><span class="st">"character"</span>)[<span class="dv">2</span>]</span>
<span id="annotated-cell-7-4"><a href="#annotated-cell-7-4"></a>            }) <span class="sc">|&gt;</span></span>
<span id="annotated-cell-7-5"><a href="#annotated-cell-7-5"></a>            <span class="fu">reduce</span>(c)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-7" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-7" data-code-lines="1" data-code-annotation="1">Passamos nosso vetor de <code>filenames</code> para a funçào <code>map</code></span>
</dd>
<dt data-target-cell="annotated-cell-7" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-7" data-code-lines="2" data-code-annotation="2"><code>(\x){}</code>: este trecho cria uma <strong>função anônima</strong> dentro do map. Isso faz com que a cada rodada da iteração <code>x</code> seja substituído pelo valor correspondente no na nossa lista de arquivos e as operações entre <code>{}</code> sejam executadas.</span>
</dd>
<dt data-target-cell="annotated-cell-7" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-7" data-code-lines="3" data-code-annotation="3">Aqui usamos a função <code>scan</code> para ler o arquivo:</span>
</dd>
</dl>
</div>
</div>
<ul>
<li><code>x</code> : é o arquivo atual na lista de arquivos passadas para o map</li>
<li><code>skip</code>: é utilizado para pular as 11 linhas iniciais do arquivo, <a href="#explorando-um-final-report">lembram delas?</a></li>
<li><code>nlines</code>: informamos que só queremos ler 1 linha do arquivo</li>
<li><code>[2]</code>: nosso arquivo possui 5 colunas, com a notação de colchetes estamos extraindo o valor da segunda coluna que contém o ID do animal</li>
</ul>
<ol start="4" type="1">
<li><code>reduce(c)</code>: como a função map retorna uma lista de ids, aqui usamos <code>reduce</code> para transformar a lista em um vetor com a função <code>c()</code> de concatenação</li>
</ol>
<p>Agora já temos tudo que precisamos para gerar o arquivo de genótipos!</p>
</section>
<section id="gerando-o-arquivo-de-genótipos" class="level3" data-number="3.4">
<h3 data-number="3.4" class="anchored" data-anchor-id="gerando-o-arquivo-de-genótipos"><span class="header-section-number">3.4</span> Gerando o arquivo de genótipos</h3>
<p>A função <code>read.snps.long</code> é a responsável por receber o vetor de <em>Final Reports</em>, fazer a leitura e organizar o nosso arquivo de genótipos.</p>
<div class="cell">
<div class="sourceCode cell-code" id="annotated-cell-8"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-8" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-8-1" class="code-annotation-target"><a href="#annotated-cell-8-1"></a>genodat<span class="ot">&lt;-</span><span class="fu">read.snps.long</span>(<span class="at">file=</span>filenames,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-8" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-8-2" class="code-annotation-target"><a href="#annotated-cell-8-2"></a>                        <span class="at">sample.id=</span>animids,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-8" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-8-3" class="code-annotation-target"><a href="#annotated-cell-8-3"></a>                        <span class="at">snp.id=</span>snpids,</span>
<span id="annotated-cell-8-4"><a href="#annotated-cell-8-4"></a>                        <span class="at">diploid=</span><span class="cn">NULL</span>, </span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-8" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-8-5" class="code-annotation-target"><a href="#annotated-cell-8-5"></a>                        <span class="at">fields=</span><span class="fu">c</span>(<span class="at">sample=</span><span class="dv">2</span>,<span class="at">snp=</span><span class="dv">1</span>,<span class="at">allele1=</span><span class="dv">3</span>,<span class="at">allele2=</span><span class="dv">4</span>,<span class="at">confidence=</span><span class="dv">5</span>),</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-8" data-target-annotation="5" onclick="event.preventDefault();">5</a><span id="annotated-cell-8-6" class="code-annotation-target"><a href="#annotated-cell-8-6"></a>                        <span class="at">codes=</span><span class="fu">c</span>(<span class="st">"A"</span>,<span class="st">"B"</span>),</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-8" data-target-annotation="6" onclick="event.preventDefault();">6</a><span id="annotated-cell-8-7" class="code-annotation-target"><a href="#annotated-cell-8-7"></a>                        <span class="at">threshold=</span><span class="fl">0.5</span>,</span>
<span id="annotated-cell-8-8"><a href="#annotated-cell-8-8"></a>                        <span class="at">lower=</span><span class="cn">TRUE</span>, </span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-8" data-target-annotation="7" onclick="event.preventDefault();">7</a><span id="annotated-cell-8-9" class="code-annotation-target"><a href="#annotated-cell-8-9"></a>                        <span class="at">skip=</span><span class="dv">11</span>,</span>
<span id="annotated-cell-8-10"><a href="#annotated-cell-8-10"></a>                        <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,</span>
<span id="annotated-cell-8-11"><a href="#annotated-cell-8-11"></a>                        <span class="at">verbose=</span><span class="cn">TRUE</span>, </span>
<span id="annotated-cell-8-12"><a href="#annotated-cell-8-12"></a>                        <span class="at">in.order=</span><span class="cn">FALSE</span>,</span>
<span id="annotated-cell-8-13"><a href="#annotated-cell-8-13"></a>                        <span class="at">every=</span><span class="dv">10000</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-8" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-8" data-code-lines="1" data-code-annotation="1">Informamos o vetor de <code>filenames</code></span>
</dd>
<dt data-target-cell="annotated-cell-8" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-8" data-code-lines="2" data-code-annotation="2">O vetor de ID dos animais <code>animids</code></span>
</dd>
<dt data-target-cell="annotated-cell-8" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-8" data-code-lines="3" data-code-annotation="3">O vetor de ID dos SNPs <code>snpids</code></span>
</dd>
<dt data-target-cell="annotated-cell-8" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-8" data-code-lines="5" data-code-annotation="4">Aqui informamos pra função em quais colunas do <em>Final Report</em> ela vai encontrar as informações referentes a amostra (<em>sample</em>), SNP, alelos ( <em>allele1</em> e <em>allele2</em> ) e o GC Score (<em>confidence</em>).</span>
</dd>
<dt data-target-cell="annotated-cell-8" data-target-annotation="5">5</dt>
<dd>
<span data-code-cell="annotated-cell-8" data-code-lines="6" data-code-annotation="5">Aqui informados de que maneira os genótipos estão codificados no <em>Final Report</em>.</span>
</dd>
<dt data-target-cell="annotated-cell-8" data-target-annotation="6">6</dt>
<dd>
<span data-code-cell="annotated-cell-8" data-code-lines="7" data-code-annotation="6">Definimos o valor mínimo de qualidade por amostra (<em>GC Score</em>)</span>
</dd>
<dt data-target-cell="annotated-cell-8" data-target-annotation="7">7</dt>
<dd>
<span data-code-cell="annotated-cell-8" data-code-lines="9" data-code-annotation="7">Ignoramos as 11 primeiras linhas do <em>Final Report</em></span>
</dd>
</dl>
</div>
</div>
<div class="sourceCode" id="annotated-cell-9"><pre class="sourceCode numberSource code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode"><a class="code-annotation-anchor" data-target-cell="annotated-cell-9" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-9-1" class="code-annotation-target"><a href="#annotated-cell-9-1"></a>Reading one call per input line</span>
<span id="annotated-cell-9-2"><a href="#annotated-cell-9-2"></a>   Sample id is in field 2</span>
<span id="annotated-cell-9-3"><a href="#annotated-cell-9-3"></a>   SNP id is in field 1</span>
<span id="annotated-cell-9-4"><a href="#annotated-cell-9-4"></a>   Allele 1 is in field 3</span>
<span id="annotated-cell-9-5"><a href="#annotated-cell-9-5"></a>   Allele 2 is in field 4</span>
<span id="annotated-cell-9-6"><a href="#annotated-cell-9-6"></a>   Confidence score is in field 5</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-9" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-9-7" class="code-annotation-target"><a href="#annotated-cell-9-7"></a>Reading SnpMatrix with 20 rows and 57636 columns</span>
<span id="annotated-cell-9-8"><a href="#annotated-cell-9-8"></a>                             Cumulative totals</span>
<span id="annotated-cell-9-9"><a href="#annotated-cell-9-9"></a>                    -----------------------------------</span>
<span id="annotated-cell-9-10"><a href="#annotated-cell-9-10"></a>    File     Line   Accepted Rejected  No call  Skipped    File name</span>
<span id="annotated-cell-9-11"><a href="#annotated-cell-9-11"></a>      20    57647    1100749    51957       14      220    ..._Report_Chk_9.txt</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-9" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-9-12" class="code-annotation-target"><a href="#annotated-cell-9-12"></a>1100749 genotypes successfully read</span>
<span id="annotated-cell-9-13"><a href="#annotated-cell-9-13"></a>51957 genotypes were rejected due to low confidence</span>
<span id="annotated-cell-9-14"><a href="#annotated-cell-9-14"></a>14 genotypes were not called</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-9" data-target-annotation="4" onclick="event.preventDefault();">4</a><span id="annotated-cell-9-15" class="code-annotation-target"><a href="#annotated-cell-9-15"></a>Warning message:</span>
<span id="annotated-cell-9-16"><a href="#annotated-cell-9-16"></a>In read.snps.long(file = filenames, sample.id = animids, snp.id = snpids,  :</span>
<span id="annotated-cell-9-17"><a href="#annotated-cell-9-17"></a>  220 lines of input file(s) were skipped</span>
<span id="annotated-cell-9-18"><a href="#annotated-cell-9-18"></a>  </span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-9" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-9" data-code-lines="1,2,3,4,5,6" data-code-annotation="1">Nos informa quais colunas foram utilizadas para leitura</span>
</dd>
<dt data-target-cell="annotated-cell-9" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-9" data-code-lines="7" data-code-annotation="2">Nos diz quantas linhas e colunas o arquivo de gentótipo possui na primeira passagem.</span>
</dd>
<dt data-target-cell="annotated-cell-9" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-9" data-code-lines="12,13,14" data-code-annotation="3">Exibe estatísticas da leitura, como o número de genótipos lidos e rejeitados.</span>
</dd>
<dt data-target-cell="annotated-cell-9" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-9" data-code-lines="15" data-code-annotation="4">Nos informa o total de linhas que foram ignoradas na leitura.</span>
</dd>
</dl>
</section>
<section id="explorando-o-objeto-de-genótipos-genodat" class="level3" data-number="3.5">
<h3 data-number="3.5" class="anchored" data-anchor-id="explorando-o-objeto-de-genótipos-genodat"><span class="header-section-number">3.5</span> Explorando o objeto de genótipos <code>genodat</code></h3>
<p>Quantos indivíduos e marcadores temos?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="fu">dim</span>(genodat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]    20 57636</code></pre>
</div>
</div>
<p>O que encontramos nas linhas?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="fu">rownames</span>(genodat) <span class="sc">|&gt;</span></span>
<span id="cb8-2"><a href="#cb8-2"></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "203900190022_R01C01" "203900190022_R05C02" "203900190022_R06C01"
[4] "203900190022_R06C02" "203900190025_R01C01" "203900190025_R01C02"</code></pre>
</div>
</div>
<p>E nas colunas?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="fu">colnames</span>(genodat) <span class="sc">|&gt;</span></span>
<span id="cb10-2"><a href="#cb10-2"></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "10080_COIII"   "10668_COIII"   "12187_ND4"     "15466_CYTB_MT"
[5] "16262_ND6"     "16370_ND6"    </code></pre>
</div>
</div>
</section>
</section>
<section id="salvando-o-arquivo-de-genótipos" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="salvando-o-arquivo-de-genótipos"><span class="header-section-number">4</span> Salvando o arquivo de genótipos</h2>
<p>Podemos agora salvar o nosso arquivo de genótipo para utilizar em análises daqui pra frente. Podemos salvá-lo tanto em formato de texto quanto no formato <code>Rdata</code>, um formato específico do R e otimizado para salvar objetos que podem ser recuperados facilmente em qualquer sessão do R.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Lembrete!
</div>
</div>
<div class="callout-body-container callout-body">
<p>Lembre que se você não especificar o caminho o arquivo será salvo na sua pasta raiz do projeto, aquela pasta que contém o .Rproj. Se desejar salvar em algum outro subdiretório basta informar o caminho no momento de salvar o arquivo.</p>
<p>Nos exemplos abaixo estamos salvando na subpasta <code>data/01_snpStats</code></p>
</div>
</div>
<section id="salvando-.rdata" class="level3" data-number="4.1">
<h3 data-number="4.1" class="anchored" data-anchor-id="salvando-.rdata"><span class="header-section-number">4.1</span> Salvando .Rdata</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="fu">save</span>(genodat,<span class="at">file=</span><span class="st">"data/01_snpStats/genodat.Rdata"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="salvando-em-formato-snpmatrix" class="level3" data-number="4.2">
<h3 data-number="4.2" class="anchored" data-anchor-id="salvando-em-formato-snpmatrix"><span class="header-section-number">4.2</span> Salvando em formato SNPMatrix</h3>
<div class="cell">
<div class="sourceCode cell-code" id="annotated-cell-14"><pre class="sourceCode numberSource r code-annotation-code number-lines code-with-copy code-annotated"><code class="sourceCode r"><a class="code-annotation-anchor" data-target-cell="annotated-cell-14" data-target-annotation="1" onclick="event.preventDefault();">1</a><span id="annotated-cell-14-1" class="code-annotation-target"><a href="#annotated-cell-14-1"></a>sampleIDs<span class="ot">&lt;-</span><span class="fu">sprintf</span>(<span class="st">'%-20s'</span>,<span class="fu">rownames</span>(genodat))</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-14" data-target-annotation="2" onclick="event.preventDefault();">2</a><span id="annotated-cell-14-2" class="code-annotation-target"><a href="#annotated-cell-14-2"></a><span class="fu">rownames</span>(genodat)<span class="ot">&lt;-</span>sampleIDs</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-14" data-target-annotation="3" onclick="event.preventDefault();">3</a><span id="annotated-cell-14-3" class="code-annotation-target"><a href="#annotated-cell-14-3"></a><span class="fu">write.SnpMatrix</span>(genodat,</span>
<span id="annotated-cell-14-4"><a href="#annotated-cell-14-4"></a>                <span class="at">file=</span><span class="st">"data/01_snpStats/genodat.txt"</span>,</span>
<span id="annotated-cell-14-5"><a href="#annotated-cell-14-5"></a>                <span class="at">quote=</span>F,</span>
<span id="annotated-cell-14-6"><a href="#annotated-cell-14-6"></a>                <span class="at">row.names=</span>T,</span>
<span id="annotated-cell-14-7"><a href="#annotated-cell-14-7"></a>                <span class="at">col.names=</span>F,</span>
<span id="annotated-cell-14-8"><a href="#annotated-cell-14-8"></a>                <span class="at">na=</span><span class="dv">5</span>,</span>
<span id="annotated-cell-14-9"><a href="#annotated-cell-14-9"></a>                <span class="at">sep=</span><span class="st">""</span>) </span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-14" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-14" data-code-lines="1" data-code-annotation="1">Aqui usamos a função <code>sprintf</code> para formatar nossos nomes de IDs para que todos tenham 20 caracteres de comprimento.</span>
</dd>
<dt data-target-cell="annotated-cell-14" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-14" data-code-lines="2" data-code-annotation="2">Então atualizamos os nomes das linhas do nosso arquivo <code>genodat</code> com os IDs formatados.</span>
</dd>
<dt data-target-cell="annotated-cell-14" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-14" data-code-lines="3" data-code-annotation="3">Por fim, salvamos o nosso arquivo em formato SNPMatrix. Observem que neste formato nossas colunas de genótipo não possuem separação entre elas. <code>sep=""</code>.</span>
</dd>
</dl>
</div>
</div>
</section>
<section id="explorando-nosso-arquivo-genodat.txt" class="level3" data-number="4.3">
<h3 data-number="4.3" class="anchored" data-anchor-id="explorando-nosso-arquivo-genodat.txt"><span class="header-section-number">4.3</span> Explorando nosso arquivo <code>genodat.txt</code></h3>
<p>E essa é a cara do nosso arquivo de genótipos.</p>
<pre><code>203900190022_R01C01 222522022120152111112111101100 ...
203900190022_R05C02 222522022101051222222111201120 ...
203900190022_R06C01 222522022111051212112111022200 ...
203900190022_R06C02 222522022220052221212021112210 ...
203900190025_R01C01 222522021021150221122201112110 ...
203900190025_R01C02 020500002020150211122000202110 ...
203900190025_R02C01 222522022112150211122000102110 ...
203900190025_R02C02 222522021121150200122002112210 ...
203900190025_R03C01 222522022021150200022100112000 ...
203900190025_R03C02 222522022221150221122000111110 ...
</code></pre>
<p>Neste formato os genótipos estão representados na forma de <strong>dosagem de alelos</strong>.</p>
<div class="grid">
<div class="g-col-3">
<table class="caption-top table">
<thead>
<tr class="header">
<th>Genótipo</th>
<th>Código</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>AA</td>
<td>2</td>
</tr>
<tr class="even">
<td>AB</td>
<td>1</td>
</tr>
<tr class="odd">
<td>BB</td>
<td>0</td>
</tr>
<tr class="even">
<td>Missing</td>
<td>5</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
</section>
<section id="código-completo" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="código-completo"><span class="header-section-number">5</span> Código Completo</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="co"># Instalando os pacotes necessários</span></span>
<span id="cb14-2"><a href="#cb14-2"></a></span>
<span id="cb14-3"><a href="#cb14-3"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb14-4"><a href="#cb14-4"></a>  <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb14-5"><a href="#cb14-5"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"snpStats"</span>)</span>
<span id="cb14-6"><a href="#cb14-6"></a></span>
<span id="cb14-7"><a href="#cb14-7"></a><span class="co"># Carregando pacotes</span></span>
<span id="cb14-8"><a href="#cb14-8"></a></span>
<span id="cb14-9"><a href="#cb14-9"></a><span class="fu">library</span>(snpStats)</span>
<span id="cb14-10"><a href="#cb14-10"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb14-11"><a href="#cb14-11"></a><span class="fu">library</span>(data.table)</span>
<span id="cb14-12"><a href="#cb14-12"></a></span>
<span id="cb14-13"><a href="#cb14-13"></a><span class="co"># Gerando vetor de Final Reports</span></span>
<span id="cb14-14"><a href="#cb14-14"></a></span>
<span id="cb14-15"><a href="#cb14-15"></a>filenames <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="at">path =</span> <span class="st">"data/01_snpStats/"</span>,</span>
<span id="cb14-16"><a href="#cb14-16"></a>           <span class="at">pattern =</span> <span class="st">"Final_Report_.*.txt"</span>,</span>
<span id="cb14-17"><a href="#cb14-17"></a>           <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb14-18"><a href="#cb14-18"></a></span>
<span id="cb14-19"><a href="#cb14-19"></a><span class="co"># Gerando vetor de SNP ids</span></span>
<span id="cb14-20"><a href="#cb14-20"></a></span>
<span id="cb14-21"><a href="#cb14-21"></a>snpids <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"data/01_snpStats/SNP-Table-Chickens50k.txt"</span>,</span>
<span id="cb14-22"><a href="#cb14-22"></a>                     <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,</span>
<span id="cb14-23"><a href="#cb14-23"></a>                     <span class="at">header =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<span id="cb14-24"><a href="#cb14-24"></a>          <span class="fu">pull</span>(Name)</span>
<span id="cb14-25"><a href="#cb14-25"></a></span>
<span id="cb14-26"><a href="#cb14-26"></a><span class="co"># Gerando vetor de Animal ids</span></span>
<span id="cb14-27"><a href="#cb14-27"></a></span>
<span id="cb14-28"><a href="#cb14-28"></a>animids<span class="ot">&lt;-</span> filenames <span class="sc">|&gt;</span></span>
<span id="cb14-29"><a href="#cb14-29"></a>            <span class="fu">map</span>(\(x){ </span>
<span id="cb14-30"><a href="#cb14-30"></a>              <span class="fu">scan</span>(x,<span class="at">skip=</span><span class="dv">11</span>,<span class="at">nlines=</span><span class="dv">1</span>,<span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,<span class="at">what=</span><span class="st">"character"</span>)[<span class="dv">2</span>]</span>
<span id="cb14-31"><a href="#cb14-31"></a>            }) <span class="sc">|&gt;</span></span>
<span id="cb14-32"><a href="#cb14-32"></a>            <span class="fu">reduce</span>(c)</span>
<span id="cb14-33"><a href="#cb14-33"></a></span>
<span id="cb14-34"><a href="#cb14-34"></a><span class="co"># Criando o arquivo de genótipo</span></span>
<span id="cb14-35"><a href="#cb14-35"></a></span>
<span id="cb14-36"><a href="#cb14-36"></a>genodat<span class="ot">&lt;-</span><span class="fu">read.snps.long</span>(<span class="at">file=</span>filenames,</span>
<span id="cb14-37"><a href="#cb14-37"></a>                        <span class="at">sample.id=</span>animids,</span>
<span id="cb14-38"><a href="#cb14-38"></a>                        <span class="at">snp.id=</span>snpids, </span>
<span id="cb14-39"><a href="#cb14-39"></a>                        <span class="at">diploid=</span><span class="cn">NULL</span>, </span>
<span id="cb14-40"><a href="#cb14-40"></a>                        <span class="at">fields=</span><span class="fu">c</span>(<span class="at">sample=</span><span class="dv">2</span>,<span class="at">snp=</span><span class="dv">1</span>,<span class="at">allele1=</span><span class="dv">3</span>,<span class="at">allele2=</span><span class="dv">4</span>,<span class="at">confidence=</span><span class="dv">5</span>),</span>
<span id="cb14-41"><a href="#cb14-41"></a>                        <span class="at">codes=</span><span class="fu">c</span>(<span class="st">"A"</span>,<span class="st">"B"</span>),</span>
<span id="cb14-42"><a href="#cb14-42"></a>                        <span class="at">threshold=</span><span class="fl">0.5</span>,</span>
<span id="cb14-43"><a href="#cb14-43"></a>                        <span class="at">lower=</span><span class="cn">TRUE</span>, </span>
<span id="cb14-44"><a href="#cb14-44"></a>                        <span class="at">skip=</span><span class="dv">11</span>,</span>
<span id="cb14-45"><a href="#cb14-45"></a>                        <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,</span>
<span id="cb14-46"><a href="#cb14-46"></a>                        <span class="at">verbose=</span><span class="cn">TRUE</span>, </span>
<span id="cb14-47"><a href="#cb14-47"></a>                        <span class="at">in.order=</span><span class="cn">FALSE</span>,</span>
<span id="cb14-48"><a href="#cb14-48"></a>                        <span class="at">every=</span><span class="dv">10000</span>)</span>
<span id="cb14-49"><a href="#cb14-49"></a></span>
<span id="cb14-50"><a href="#cb14-50"></a><span class="co"># Salvando em formato Rdata</span></span>
<span id="cb14-51"><a href="#cb14-51"></a></span>
<span id="cb14-52"><a href="#cb14-52"></a><span class="fu">save</span>(genodat,<span class="at">file=</span><span class="st">"data/01_snpStats/genodat.Rdata"</span>)</span>
<span id="cb14-53"><a href="#cb14-53"></a></span>
<span id="cb14-54"><a href="#cb14-54"></a></span>
<span id="cb14-55"><a href="#cb14-55"></a><span class="co"># Salvando em formato SNPMatrix</span></span>
<span id="cb14-56"><a href="#cb14-56"></a></span>
<span id="cb14-57"><a href="#cb14-57"></a>sampleIDs<span class="ot">&lt;-</span><span class="fu">sprintf</span>(<span class="st">'%-20s'</span>,<span class="fu">rownames</span>(genodat))</span>
<span id="cb14-58"><a href="#cb14-58"></a><span class="fu">rownames</span>(genodat)<span class="ot">&lt;-</span>sampleIDs  </span>
<span id="cb14-59"><a href="#cb14-59"></a><span class="fu">write.SnpMatrix</span>(genodat, </span>
<span id="cb14-60"><a href="#cb14-60"></a>                <span class="at">file=</span><span class="st">"data/01_snpStats/genodat.txt"</span>,</span>
<span id="cb14-61"><a href="#cb14-61"></a>                <span class="at">quote=</span>F,</span>
<span id="cb14-62"><a href="#cb14-62"></a>                <span class="at">row.names=</span>T,</span>
<span id="cb14-63"><a href="#cb14-63"></a>                <span class="at">col.names=</span>F,</span>
<span id="cb14-64"><a href="#cb14-64"></a>                <span class="at">na=</span><span class="dv">5</span>,</span>
<span id="cb14-65"><a href="#cb14-65"></a>                <span class="at">sep=</span><span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/paulobarros\.github\.io\/intro-stats-gen\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./index.html" class="pagination-link" aria-label="Bem-vind@s!">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Bem-vind@s!</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p><img src="images/MM.png" class="img-fluid" width="45"></p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    <div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/paulobarros/intro-stats-gen/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
<p>© CC-By Paulo Barros, 2024</p>
</div>
  </div>
</footer>




</body></html>